### ISSUE TITLE
Rename error when crossing devices (invalid cross-device link)

### ISSUE THREAD

**Issue by jradwan** [jradwan/Piwigo-physical_photo_move#8]
When moving an item from one physical folder to another, if the source and target folders are on different devices (at the OS level), the PHP rename command will throw a warning.

For photos:

`Warning: rename(./galleries/places/batch-test-02.png,./galleries/exp_PG_test/batch-test-02.png): Operation not permitted in /usr/local/piwigo-beta/plugins/physical_photo_move/include/functions.inc.php on line 187`

For albums:

```
Warning: rename(): The first argument to copy() function cannot be a directory in /usr/local/piwigo-beta/plugins/physical_photo_move/include/functions.inc.php on line 435

Warning: rename(./galleries/exp_PG_test,./galleries/places/exp_PG_test): Invalid cross-device link in /usr/local/piwigo-beta/plugins/physical_photo_move/include/functions.inc.php on line 435

Fatal error: Uncaught UnexpectedValueException: DirectoryIterator::__construct(./galleries/places/exp_PG_test): failed to open dir: No such file or directory in /usr/local/piwigo-beta/plugins/physical_photo_move/include/functions.inc.php:519 Stack trace: #0 /usr/local/piwigo-beta/plugins/physical_photo_move/include/functions.inc.php(519): DirectoryIterator->__construct('./galleries/pla...') #1 /usr/local/piwigo-beta/plugins/physical_photo_move/include/functions.inc.php(436): ppm_chmod_r('./galleries/pla...') #2 /usr/local/piwigo-beta/plugins/physical_photo_move/include/functions.inc.php(82): ppm_move_album('7', '37', false) #3 /usr/local/piwigo-beta/plugins/physical_photo_move/admin.php(69): ppm_move_item('7', '37', false, 'album') #4 /usr/local/piwigo-beta/admin/plugin.php(68): include_once('/usr/local/piwi...') #5 /usr/local/piwigo-beta/admin.php(311): include('/usr/local/piwi...') #6 {main} thrown in /usr/local/piwigo-beta/plugins/physical_photo_move/include/functions.inc.php on line 519
```

In the case of photos, it doesn't seem to stop processing and the files are correctly moved. For albums, however, the fatal error prevents the move from happening and the code from continuing.

This really only seems to happen if you use symbolic links, which I do: under Piwigo's galleries folder (i.e., /usr/local/piwigo/galleries) I have symbolic links out to my NAS. 

It looks like maybe I should use the copy() and unlink() functions in this case.

**Comment by RStoroy** [comment]
Is this issue ever gonna be fixed?

**Comment by jradwan** [comment]
I haven't worked on it (sorry!) for no reason other than being selfish (I don't typically run into it, so it doesn't affect me and this isn't a high priority).

I will be doing some brief testing of this extension against the latest version 13 beta of Piwigo ... if I have time I might look into this.

See also https://bugs.php.net/bug.php?id=54097.

**Comment by jradwan** [comment]
@RStoroy I took a shot at finally fixing this. If you have a test Piwigo site you'd like to try this on, just grab the updated include/functions.inc.php file here:

https://github.com/jradwan/Piwigo-physical_photo_move/blob/2.11/include/functions.inc.php

And replace it with the one in the plug-in folder on your server (piwigo/plugins/physical_photo_move/include/). Make a backup copy first just in case so you can restore the old version if necessary. I've done a bunch of testing on my own test site (moving individual photos, albums, videos, etc.) across devices (via symlinks) and it seems to work as expected. Just making sure I didn't introduce any other regression issues by changing how the "move" works (switching from PHP's rename function to using copy, unlink (delete file) and rmdir (delete folder).

If everything looks good, I'll incorporate this into the next release (2.11) along with making sure everything works with Piwigo 13.

Thanks.

**Comment by RStoroy** [comment]
Many thanks for your work on this. I will test i soon Best Regards Robert

**Comment by jradwan** [comment]
Fixed in v2.20.
